### PR TITLE
[Datasets] Add `SimpleTorchDatasource`

### PR DESCRIPTION
**Pull request by bveeramani**
## Why are these changes needed?

It's difficult to use `torchvision` datasets with Ray ML. This PR makes it easier to use Torch datasets with Ray Data.

**Before**:
```
class CIFAR10Datasource(FileBasedDatasource):
    def _read_file(self, f: "pyarrow.NativeFile", path: str, **reader_args):
        data = pickle.load(io.BytesIO(f.readall()), encoding="bytes")

        features: np.ndarray = data[b"data"]
        num_features = features.shape[1]
        columns = list(map(str, range(num_features)))
        df = pd.DataFrame(features, columns=columns)

        labels: List[int] = data[b"labels"]
        df["label"] = labels

        return df

    def _rows_per_file(self):
        return 1

paths =  [os.path.join("data", f"data_batch_{i}") for i in range(1, 6)]
ds = ray.data.read_datasource(CIFAR10Datasource(), paths=paths)
```

**After:**
``` 
dataset_factory = lambda: return torchvision.datasets.CIFAR10("data")
ds = ray.data.read_datasource(
    TorchDatasource(), parallelism=1, dataset_factory=dataset_factory
)
```

## Related issue number

Towards https://github.com/ray-project/ray/issues/23977

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/datasource/torch_datasource.py
Comment (ericl) [inline python/ray/data/datasource/torch_datasource.py:12]:
```suggestion
class SimpleTorchDatasource(Datasource[T]):
```

This clarifies that it's not a scalable datasource.

File: python/ray/data/datasource/torch_datasource.py
Comment (ericl) [inline python/ray/data/datasource/torch_datasource.py:22]: Should dataset_factory be passed as a constructor arg to TorchDatasource?

Comment (bveeramani) [reply python/ray/data/datasource/torch_datasource.py:22]: Discussed with @clarkzinzow, and we decided to place `dataset_factory` in `prepare_read` to be consistent with the other datasources.

File: python/ray/data/datasource/torch_datasource.py
Comment (bveeramani) [inline python/ray/data/datasource/torch_datasource.py:12]: Not sure where to place this class in the documentation (if at all). From what I can tell, other datasources aren't listed in the API reference. Would appreciate thoughts on this.

Comment (bveeramani) [reply python/ray/data/datasource/torch_datasource.py]: Fixed

File: python/ray/data/datasource/torch_datasource.py
Comment (clarkzinzow) [inline python/ray/data/datasource/torch_datasource.py:30]: Specify the type as a string so it's not a hard dependency for all Datasets users.
```suggestion
        self, parallelism: int, dataset_factory: Callable[[], "torch.utils.data.Dataset"]
```

File: python/ray/data/datasource/torch_datasource.py
Comment (clarkzinzow) [inline python/ray/data/datasource/torch_datasource.py:44]: Good check!

File: python/ray/data/datasource/torch_datasource.py
Comment (clarkzinzow) [inline python/ray/data/datasource/torch_datasource.py:52]: Should return an iterator rather than a concrete list.
```suggestion
        def read_fn() -> Iterator[Block]:
```

File: python/ray/data/datasource/torch_datasource.py
Comment (clarkzinzow) [inline python/ray/data/datasource/torch_datasource.py:56]:
```suggestion
            yield block
```

File: python/ray/data/tests/test_dataset_formats.py
Comment (clarkzinzow) [inline python/ray/data/tests/test_dataset_formats.py:1476]: By explicitly setting the `batch_size` to `None` you disable batching, so this will return the entire block!
```suggestion
    actual_data = list(next(ray_dataset.iter_batches(batch_size=None)))
```

Comment (clarkzinzow) [reply python/ray/data/tests/test_dataset_formats.py:1476]: Another option would be checking the number of blocks and the content separately:
```python
ray_dataset = ray_dataset.fully_executed()
assert ray_dataset.num_blocks() == 1
assert ray_dataset.take_all() == expected_data
```

File: python/ray/data/datasource/torch_datasource.py
Comment (clarkzinzow) [inline python/ray/data/datasource/torch_datasource.py:50]: Nice warning!

File: python/ray/data/tests/test_dataset_formats.py
Comment (clarkzinzow) [inline python/ray/data/tests/test_dataset_formats.py:1478]: Could you also check that passing an instantiated Torch dataset raises an error? E.g.
```python
with pytest.raises(ValueError):
    ray.data.read_datasource(
        SimpleTorchDatasource,
        parallelism=1,
        dataset_factory=dataset_factory(),
    )
```

**Review by clarkzinzow — COMMENTED** [review]
Looking great overall! Just a few small items.

File: python/ray/data/datasource/torch_datasource.py
Comment (bveeramani) [reply python/ray/data/datasource/torch_datasource.py]: Fixed

File: python/ray/data/datasource/torch_datasource.py
Comment (bveeramani) [reply python/ray/data/datasource/torch_datasource.py]: Fixed

File: python/ray/data/datasource/torch_datasource.py
Comment (bveeramani) [reply python/ray/data/datasource/torch_datasource.py]: Fixed

File: python/ray/data/tests/test_dataset_formats.py
Comment (bveeramani) [reply python/ray/data/tests/test_dataset_formats.py:1476]: Fixed

File: python/ray/data/tests/test_dataset_formats.py
Comment (bveeramani) [reply python/ray/data/tests/test_dataset_formats.py:1478]: Fixed

File: python/ray/data/datasource/__init__.py
Comment (matthewdeng) [inline python/ray/data/datasource/__init__.py:27]: Are we okay with having this defined generically in `ray.data` as opposed to having it be an application-specific Datasource that exists in `ray.air` (`ray.ml`)? @clarkzinzow @ericl

File: python/ray/data/datasource/torch_datasource.py
Comment (matthewdeng) [inline python/ray/data/datasource/torch_datasource.py:26]: nit: If you want this to be as minimal as possible you can use `take`.
```suggestion
        >>> dataset.take(1)
        (<PIL.Image.Image image mode=L size=28x28 at 0x1142CCA60>, 5)
```

File: python/ray/data/datasource/torch_datasource.py
Comment (matthewdeng) [inline python/ray/data/datasource/torch_datasource.py:38]:
```suggestion
            dataset_factory: A no-argument function that returns the Torch dataset to be read.
```

File: python/ray/data/tests/test_dataset_formats.py
Comment (matthewdeng) [inline python/ray/data/tests/test_dataset_formats.py:1489]: Test is failing, not sure if this is the cause...
```suggestion
            SimpleTorchDatasource(),
```

**Review by matthewdeng — COMMENTED** [review]
Looks great!

File: python/ray/data/tests/test_dataset_formats.py
Comment (bveeramani) [reply python/ray/data/tests/test_dataset_formats.py:1489]: Fixed

File: python/ray/data/datasource/__init__.py
Comment (clarkzinzow) [reply python/ray/data/datasource/__init__.py:27]: @matthewdeng That's a good point! My vote would be having it defined in `ray.data` for now, with the rest of our datasources. Since the I/O APIs are still exposed via `ray.data`, I'd like to keep the datasources we've defined in `ray.data` for ease-of-use in our "read --> preprocess --> train" getting started flow, where we delay interaction with the AIR package until we get to the "preprocess" stage.

I don't feel too strongly about this and haven't thought through it thoroughly, so I'm happy to hear other thoughts here.

**Review by clarkzinzow — APPROVED** [review]
LGTM, nice work!

File: python/ray/data/datasource/__init__.py
Comment (matthewdeng) [reply python/ray/data/datasource/__init__.py:27]: SGTM! The original concern was that these datasources would be not general enough to want to expose to users, but I think the documentation and having them prefixed with `Simple` should address any misunderstandings.

But on that note, should these be added to the Ray docs somewhere? I saw API docs for `Datasource` but not any of the implementations.

**Comment by clarkzinzow** [comment]
Ah @bveeramani just saw your comment on documentation, we should:
1. Expose the `Datasource` implementations in the API docs (i.e. adding to [here](https://github.com/ray-project/ray/blob/master/doc/source/data/package-ref.rst)).
2. Add the `SimpleTorchDatasource` to those API docs.
3. Add example usage of this datasource to the ["Creating Datasets" user guide](https://docs.ray.io/en/master/data/creating-datasets.html).

Would you be ok with adding these documentation updates in this PR?

File: python/ray/data/datasource/__init__.py
Comment (clarkzinzow) [reply python/ray/data/datasource/__init__.py:27]: >But on that note, should these be added to the Ray docs somewhere? I saw API docs for Datasource but not any of the implementations.

@matthewdeng Yep I noticed that too, left a [suggestion](https://github.com/ray-project/ray/pull/23926#issuecomment-1105534061) for how to add this datasource (and others) to the docs!

File: python/ray/data/datasource/__init__.py
Comment (bveeramani) [inline python/ray/data/datasource/__init__.py:1]: Ended up running `isort` on the file since it was disorganized.

File: doc/source/data/creating-datasets.rst
Comment (clarkzinzow) [inline doc/source/data/creating-datasets.rst:73]:
```suggestion
If you already have a Torch dataset available, you can create a Ray Dataset using 
```

**Review by clarkzinzow — APPROVED** [review]
LGTM, thank you for making the docs changes!

File: python/ray/data/datasource/torch_datasource.py
Comment (clarkzinzow) [inline python/ray/data/datasource/torch_datasource.py:28]: Looks like this example is failing in CI, could you add `# doctest: +SKIP` to all but the import lines? E.g. https://github.com/ray-project/ray/blob/cadbd1fb4e2a19e5e8bbd0e9bca9c673f8d0856c/python/ray/data/datasource/parquet_datasource.py#L168-L173

Comment (bveeramani) [reply python/ray/data/datasource/torch_datasource.py]: Fixed

**Comment by clarkzinzow** [comment]
@bveeramani Could you merge latest master? It looks like a major release of `cryptography` [broke](https://github.com/ray-project/ray/commit/d9d9fbb0441bf52c8bbd53bf67011ec3f77f5697) our CI, so none of the Datasets tests ran in the last CI job.

The PR LGTM, so after the next full CI run passes, we can merge!

**Review by clarkzinzow — APPROVED** [review]
Awesome, great work! 👏

**Pull request merged**
